### PR TITLE
feat: add FunctionalTableTransformer

### LINKED ISSUE #858: Functional table transformer
### Is your feature request related to a problem?

Currently, users can transform a table either with a `TableTransformer` or with the methods of `Table`. However, only a `TableTransformer` can be used in a `SequentialTableTransformer` (see #802).

### Desired solution

* Add a new class `FunctionalTableTransformer`
* **Superclass:** `TableTransformer`
* **Constructor parameters:**
  * `transformer: Callable[[Table], Table]` The function transforms a table and returns the transformed table.
* `is_fitted`: Always `True`, immediately after constructing the model.
* `fit`: Is a no-OP. Just return `self`.
* `transform`: Call the `transformer` you got in the constructor and return the result.

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by Tarmandan**
Closes #858 

### Summary of Changes

This implements a Transformer that wraps a Callable[[Table], Table] so operations on Tables can be inserted into the operation order of a SequentialTableTransformer.
Note that this transformer inherently cannot be invertible.
Since there is no type checking at runtime, callables with wrong argument or return types will only throw an exception upon calling transform but not during init.

### CODE REVIEW

**Comment by Tarmandan** [comment]
The linter found 2 problems. I fixed the first by adding type hint to the funciton valid_callable in the tests.
However the other problem is that the linter correctly notices that invalid_callable has the wrong types. I am not sure if the test is superfluous. The test for internal error of the callable raises an exception in the same try-except, just later.

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    funct:
+        The Callable that receives a table and returns a table that is to be wrapped.
+    """
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Dunder methods
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def __init__(self,
+                 funct: Callable[[Table], Table],
```
Comment (lars-reimann) [inline src/safeds/data/tabular/transformation/_functional_table_transformer.py]:
```suggestion
                 transformer: Callable[[Table], Table],
```

See issue. Use a rename refactoring. Don't forget to update the docstring.

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                 ) -> None:
+        super().__init__(None)
+        self._func = funct
```
Comment (lars-reimann) [inline src/safeds/data/tabular/transformation/_functional_table_transformer.py:34]: Rename attribute to `self._transformer`.

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    from collections.abc import Callable
+
+    from safeds.data.tabular.containers import Table
+
+from ._table_transformer import TableTransformer
+
+
+class FunctionalTableTransformer(TableTransformer):
+    """
+    Learn a transformation for a set of columns in a `Table` and transform another `Table` with the same columns.
```
Comment (lars-reimann) [inline src/safeds/data/tabular/transformation/_functional_table_transformer.py:17]: Should be updated to better reflect the purpose of this class.

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        ------
+        Exception:
+            Raised when the wrapped callable encounters an error.
+
+        """
+        try:
+            return self._func(table)
+        except Exception as e:
+            # TODO Evaluate if switch to non-generic exception is useful, as _func can be any callable
+            raise Exception("The underlying function encountered an error") from e  # noqa: TRY002
```
Comment (lars-reimann) [inline src/safeds/data/tabular/transformation/_functional_table_transformer.py:98]: Would also be fine to remove the wrapping into another `Exception` completely and just let the original exception bubble up. This addtional exception does not add useful information.

**Review by lars-reimann — CHANGES_REQUESTED** [review]

**Comment by Tarmandan** [comment]
The test for invalid callable type has been removed.
The only problem remaining is test coverage. The __hash__ test in test_table_transformer.py should be adjusted, but FunctionalTableTransformer should only give different hash when wrapping different callable, so it will need its own tests. It might be less work to implement it in the test_functional_table_transformer.py, but would break the pattern.

**Comment by lars-reimann** [comment]
For now, it should suffice to add the transformer to https://github.com/Safe-DS/Stdlib/blob/69a780cab1a5dcab33c8c7a36f03bd2eb3367683/tests/safeds/data/tabular/transformation/test_table_transformer.py#L35. We can adjust the test later so `hash` is tested more in-depth for all transformers. Maybe add an issue for that.

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py (hunk 1 above)
Comment (Tarmandan) [reply src/safeds/data/tabular/transformation/_functional_table_transformer.py]: Done

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py (hunk 2 above)
Comment (Tarmandan) [reply src/safeds/data/tabular/transformation/_functional_table_transformer.py]: Done

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py (hunk 4 above)
Comment (Tarmandan) [reply src/safeds/data/tabular/transformation/_functional_table_transformer.py]: Wrapper has been removed

File: src/safeds/data/tabular/transformation/_functional_table_transformer.py (hunk 3 above)
Comment (Tarmandan) [reply src/safeds/data/tabular/transformation/_functional_table_transformer.py]: Text has been adjusted, please check if adequate

**Review by lars-reimann — APPROVED** [review]
Thanks, nice work.

**Pull request merged**